x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java

```diff
@@ -610,27 +612,84 @@ private static char[] getPbkdf2Hash(SecureString data, int cost, String prefix)
         }
     }
 
+    private static int parsePbkdf2Iterations(char[] hash, String prefix) {
+        int separator = -1;
+        for (int i = prefix.length(); i < hash.length; i++) {
+            if (hash[i] == '$') {
+                separator = i;
+                break;
+            }
+        }
+        if (separator == -1) {
+            throw new IllegalArgumentException("unknown hash format");
+        }
+        try {
+            return Integer.parseInt(new String(hash, prefix.length(), separator - prefix.length()));
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("hash must include a valid iteration number", e);
+        }
+    }
+
     private static boolean verifyPbkdf2Hash(SecureString data, char[] hash, String prefix) {
-        // Base64 string length : (4*(n/3)) rounded up to the next multiple of 4 because of padding.
-        // n is 32 (PBKDF2_KEY_LENGTH in bytes), so tokenLength is 44
-        final int tokenLength = 44;
+        if (CharArrays.charsBeginsWith(prefix, hash) == false) {
+            return false;
+        }
+
+        int separator1 = -1, separator2 = -1;
+        for (int i = 0; i < hash.length; i++) {
+            if (hash[i] == '$') {
+                separator1 = i;
+                break;
+            }
+        }
+        if (separator1 == -1) {
+            return false;
+        }
+        for (int i = separator1 + 1; i < hash.length; i++) {
+            if (hash[i] == '$') {
+                separator2 = i;
+                break;
+            }
+        }
+        if (separator2 == -1) {
+            return false;
+        }
```

Comment:
nit: instead of walking the char[] to find the seperators here, can we pass in the iterations and use that to determine where in the char[] the seperators live. For example, with the iterations passed in we know where the first seperator lives -> prefix.length + String.valueOf(iterations).length, we need to find the hashChars anyway so we can iterate the char[] from the hash.length backwards to the first found `$` and then the salt is squished in between like you have here. I think it avoids a few CPU cycles and does not require rediscovering information we already discovered when choosing the hash and think it may read abit cleaner. mostly just an opinion and cosmetic, so feel free to ignore.

Reply:
It's a good idea. I'll adopt it in the code.